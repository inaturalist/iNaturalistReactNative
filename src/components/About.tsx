import { useNavigation } from "@react-navigation/native";
import {
  Body1,
  Body2,
  Button,
  Heading4,
  ScrollViewWrapper,
  Tabs,
  UnderlinedLink,
  ViewWrapper
} from "components/SharedComponents";
import { Image, Pressable, View } from "components/styledComponents";
import { t } from "i18next";
import React, { useState } from "react";
import { getBuildNumber, getVersion } from "react-native-device-info";
import { useDebugMode } from "sharedHooks";

const aboutID = "about";
const teamID = "team";

const About = ( ) => {
  const navigation = useNavigation();
  const [activeTab, setActiveTab] = useState( aboutID );
  const [count, setCount] = useState( 0 );
  const appVersion = getVersion();
  const buildVersion = getBuildNumber();
  const { isDebug, toggleDebug } = useDebugMode( );

  const onTermsPressed = async () => {
    const url = "https://www.inaturalist.org/pages/terms";
    navigation.navigate( "FullPageWebView", {
      title: t( "TERMS-OF-USE" ),
      initialUrl: url,
      loggedIn: false
    } );
  };

  const onPrivacyPressed = async () => {
    const url = "https://www.inaturalist.org/pages/privacy";
    navigation.navigate( "FullPageWebView", {
      title: t( "PRIVACY-POLICY" ),
      initialUrl: url,
      loggedIn: false
    } );
  };

  const onCommunityPressed = async () => {
    const url = "https://www.inaturalist.org/pages/community+guidelines";
    navigation.navigate( "FullPageWebView", {
      title: t( "COMMUNITY-GUIDELINES" ),
      initialUrl: url,
      loggedIn: false
    } );
  };

  const onVersionPressed = () => {
    if ( ( count + 1 ) % 3 === 0 ) {
      toggleDebug();
    }
    setCount( count + 1 );
  };

  return (
    <ViewWrapper>
      <Tabs
        tabs={[
          {
            id: aboutID,
            text: t( "ABOUT" ),
            onPress: () => {
              setActiveTab( aboutID );
            }
          },
          {
            id: teamID,
            text: t( "TEAM" ),
            onPress: () => {
              setActiveTab( teamID );
            }
          }
        ]}
        activeId={activeTab}
      />
      <ScrollViewWrapper>
        {activeTab === aboutID && (
          <View className="px-4 py-8">
            <Image
              className="w-[65%] max-h-12 xl:max-h-24 2xl:max-h-32 object-center self-center mb-8"
              resizeMode="contain"
              source={require( "images/inaturalist-dark.png" )}
              accessibilityIgnoresInvertColors
            />
            <Heading4 className="mb-3">
              {t( "INATURALIST-MISSION-VISION" )}
            </Heading4>
            <Body2 className="mb-5">
              {t( "iNaturalist-mission-is-to-connect" )}
            </Body2>
            <Body2 className="mb-8">{t( "iNaturalists-vision-is-a-world" )}</Body2>
            <Heading4 className="mb-3">{t( "WHAT-IS-INATURALIST" )}</Heading4>
            <Body2 className="mb-5">
              {t( "iNaturalist-helps-you-identify" )}
            </Body2>
            <Body2 className="mb-5">{t( "Whats-more-by-recording" )}</Body2>
            <Body2 className="mb-8">{t( "iNaturalist-is-supported-by" )}</Body2>
            <UnderlinedLink
              className="mb-5"
              onPress={() => onTermsPressed()}
            >
              {t( "Terms-of-Use" )}
            </UnderlinedLink>
            <UnderlinedLink
              className="mb-5"
              onPress={() => onPrivacyPressed()}
            >
              {t( "Privacy-Policy" )}
            </UnderlinedLink>
            <UnderlinedLink
              className="mb-8"
              onPress={() => onCommunityPressed()}
            >
              {t( "Community-Guidelines" )}
            </UnderlinedLink>
            <Pressable
              accessibilityRole="button"
              className="items-center justify-center"
              onPress={() => onVersionPressed()}
            >
              <Body1>{ t( "Version-app-build", { appVersion, buildVersion } )}</Body1>
            </Pressable>
            {isDebug && (
              <Button
                text="TURN OFF DEBUG MODE"
                className="mt-5 bg-deeppink"
                level="primary"
                onPress={() => {
                  setCount( 0 );
                  toggleDebug();
                }}
              />
            )}
          </View>
        )}
        {activeTab === teamID && (
          <View className="px-4 py-8">
            <Heading4 className="mb-3">{t( "INATURALIST-TEAM" )}</Heading4>
            <Body2 className="mb-8">
              {t( "iNaturalists-apps-are-designed-and-developed-3" )}
            </Body2>
            <Heading4 className="mb-3">{t( "INATURALIST-COMMUNITY" )}</Heading4>
            <Body2 className="mb-8">
              {t( "iNaturalist-is-supported-by-our-community" )}
            </Body2>
            <Heading4 className="mb-3">{t( "INATURALIST-NETWORK" )}</Heading4>
            <Body2 className="mb-8">{t( "The-iNaturalist-Network" )}</Body2>
            <Heading4 className="mb-3">{t( "COLLABORATORS" )}</Heading4>
            <Body2>{t( "The-models-that-suggest-species" )}</Body2>
          </View>
        )}
      </ScrollViewWrapper>
    </ViewWrapper>
  );
};

export default About;
