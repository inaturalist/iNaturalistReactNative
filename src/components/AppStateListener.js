import { focusManager } from "@tanstack/react-query";
import useDeviceStorageFull from "components/Camera/hooks/useDeviceStorageFull";
import { useEffect, useState } from "react";
import { AppState } from "react-native";
import { log } from "sharedHelpers/logger";
import {
  usePerformance
} from "sharedHooks";
import { isDebugMode } from "sharedHooks/useDebugMode";

const logger = log.extend( "AppStateListener" );

const AppStateListener = ( ) => {
  const { loadTime } = usePerformance( {
    screenName: "AppStateListener"
  } );
  if ( isDebugMode( ) ) {
    logger.info( loadTime );
  }
  const { deviceStorageFull, showStorageFullAlert } = useDeviceStorageFull( );
  const [deviceStorageFullShown, setDeviceStorageFullShown] = useState( false );

  useEffect( ( ) => {
    const subscription = AppState.addEventListener( "change", nextAppState => {
      if ( nextAppState === "active" ) {
        // Check storage and show alert as needed when app becomes active
        if ( deviceStorageFull && !deviceStorageFullShown ) {
          showStorageFullAlert( );
          setDeviceStorageFullShown( true );
        }

        focusManager.setFocused( true );
      }
    } );

    return ( ) => subscription.remove( );
  }, [deviceStorageFull, deviceStorageFullShown, showStorageFullAlert] );

  return null;
};

export default AppStateListener;
