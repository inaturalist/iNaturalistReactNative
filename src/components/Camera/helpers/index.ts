import { Alert } from "react-native";
import { CameraRuntimeError } from "react-native-vision-camera";

import { log } from "../../../../react-native-logs.config";

const logger = log.extend( "CameraView" );

type Event = {
  log: string;
}

const handleClassifierError = ( error: CameraRuntimeError ) => {
  // When we hit this error, there is an error with the classifier.
  logger.error( "handleClassifierError: ", JSON.stringify( error ) );
  Alert.alert( "error", JSON.stringify( error ) );
};

const handleDeviceNotSupported = ( error: CameraRuntimeError ) => {
  // When we hit this error, something with the current device is not supported.
  Alert.alert( "error", JSON.stringify( error ) );
};

const handleCaptureError = ( error: CameraRuntimeError ) => {
  // When we hit this error, taking a photo did not work correctly
  logger.error( "handleCaptureError: ", JSON.stringify( error ) );
  Alert.alert( "error", JSON.stringify( error ) );
};

const handleCameraError = ( error: CameraRuntimeError ) => {
  // This error is thrown when it does not fit in any of the above categories.
  logger.error( "handleCameraError ", JSON.stringify( error ) );
  Alert.alert( "error", JSON.stringify( error ) );
};

const handleLog = ( event: Event ) => {
  logger.debug( event.log );
};

export {
  handleCameraError,
  handleCaptureError,
  handleClassifierError,
  handleDeviceNotSupported,
  handleLog
};
