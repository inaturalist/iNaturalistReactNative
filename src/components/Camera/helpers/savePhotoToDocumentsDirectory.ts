import {
  rotatedOriginalPhotosPath
} from "appConstants/paths";
import RNFS from "react-native-fs";
import type {
  PhotoFile
} from "react-native-vision-camera";
import resizeImage from "sharedHelpers/resizeImage";
import { unlink } from "sharedHelpers/util";

const savePhotoToDocumentsDirectory = async (
  cameraPhoto: PhotoFile
) => {
  const path = rotatedOriginalPhotosPath;
  await RNFS.mkdir( path );
  // Move the image with ImageResizer (for legacy reasons, because we
  // used to use it to rotate the photo)
  const image = await resizeImage(
    cameraPhoto.path,
    {
      width: cameraPhoto.width,
      height: cameraPhoto.height,
      outputPath: path
    }
  );
  // Remove original photo
  await unlink( cameraPhoto.path );
  return image;
};

export default savePhotoToDocumentsDirectory;
