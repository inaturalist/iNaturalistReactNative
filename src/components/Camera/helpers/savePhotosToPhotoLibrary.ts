import { CameraRoll } from "@react-native-camera-roll/camera-roll";
import {
  permissionResultFromMultiple,
  READ_WRITE_MEDIA_PERMISSIONS
} from "components/SharedComponents/PermissionGateContainer.tsx";
import { t } from "i18next";
import {
  Alert
} from "react-native";
import { checkMultiple, RESULTS } from "react-native-permissions";
import { log } from "sharedHelpers/logger";

import { displayName as appName } from "../../../../app.json";

const logger = log.extend( "savePhotosToPhotoLibrary" );

// Save URIs to camera photo library (if a photo was taken using the app,
// we want it accessible in the camera's folder, as if the user has taken those photos
// via their own camera app).
// One could argue this is a private method and shouldn't be exported and
// doesn't need to be tested... but hooks are complicated and this hook might
// be too complicated, so this at least makes it easy to test this one part
// ~~~kueda20240614
// $FlowIgnore
export async function savePhotosToPhotoLibrary(
  uris: [string],
  location: Object
) {
  const readWritePermissionResult = permissionResultFromMultiple(
    await checkMultiple( READ_WRITE_MEDIA_PERMISSIONS )
  );
  const savedPhotoUris = await uris.reduce(
    async ( memo, uri ) => {
      const savedUris = await memo;
      try {
        const saveOptions = {};
        // One quirk of CameraRoll is that if you want to write to an album, you
        // need readwrite permission, but we don't want to ask for that here
        // b/c it might come immediately after asking for *add only*
        // permission, so we're checking to see if we have that permission
        // and skipping the album if we don't
        if ( readWritePermissionResult === RESULTS.GRANTED ) {
          saveOptions.type = "photo";
          // Note: we do not translate our brand name, so this should not be
          // globalized
          saveOptions.album = appName;
        }
        if ( location ) {
          saveOptions.latitude = location.latitude;
          saveOptions.longitude = location.longitude;
          saveOptions.horizontalAccuracy = location.positional_accuracy;
        }
        const savedPhotoUri = await CameraRoll.save( uri, saveOptions );
        savedUris.push( savedPhotoUri );
        return savedUris;
      } catch ( cameraRollSaveError ) {
        // should never get here since in usePrepareStoreAndNavigate we check for device full
        // and skip saving to photo library
        if (
          cameraRollSaveError.message.match( /No space left on device/ )
          || cameraRollSaveError.message.match( /PHPhotosErrorDomain error 3305/ )
        ) {
          Alert.alert(
            t( "Not-enough-space-left-on-device" ),
            t( "Not-enough-space-left-on-device-try-again" ),
            [{ text: t( "OK" ) }]
          );
          return savedUris;
        }
        // This means an iOS user denied access
        // (https://developer.apple.com/documentation/photokit/phphotoserror/code/accessuserdenied).
        // In theory we should not even have called this function when that
        // happens, but we're still seeing this in the logs. They should be
        // prompted to grant permission the next time they try so this is
        // probably safe to ignore.
        if ( !cameraRollSaveError.message.match( /error 3311/ ) ) {
          logger.error( cameraRollSaveError );
          return savedUris;
        }
        throw cameraRollSaveError;
      }
    },
    // We need the initial value even if we're not using it, otherwise reduce
    // will treat the first item in the array as the initial value and not
    // call the reducer function on it
    Promise.resolve( [] )
  );
  return savedPhotoUris;
}

export default savePhotosToPhotoLibrary;
