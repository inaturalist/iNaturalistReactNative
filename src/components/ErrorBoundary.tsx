import {
  Body1,
  Button,
  Heading1,
  Heading2
} from "components/SharedComponents";
import {
  SafeAreaView,
  ScrollView
} from "components/styledComponents";
import { t } from "i18next";
import React from "react";
import RNRestart from "react-native-restart";
import { log } from "sharedHelpers/logger";

const logger = log.extend( "ErrorBoundary" );

interface Props {
  children: React.ReactNode;
}

interface State {
  error: Error | null;
  info: React.ErrorInfo | null;
}

// https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary
class ErrorBoundary extends React.Component<Props, State> {
  constructor( props: Props ) {
    super( props );
    this.state = { error: null, info: null };
  }

  static getDerivedStateFromError( error: Error ) {
    // Update state so the next render will show the fallback UI.
    return { error };
  }

  componentDidCatch( error: Error, info: React.ErrorInfo ) {
    this.setState( { info } );
    if ( info?.componentStack ) {
      // componentStack is a little more informative than what's generally in
      // stack, so put that in there before logging
      error.stack = info.componentStack;
    }
    logger.error( error );
  }

  render() {
    if ( this.state.error ) {
      // TODO allow this to take a fallback component as a prop so individual
      // boundaries can customize their error states
      return (
        <SafeAreaView className="flex-1">
          <ScrollView className="p-5 bg-white">
            <Heading1 className="my-3">{ t( "Something-went-wrong" ) }</Heading1>
            <Body1>
              { t( "If-youre-seeing-this-error" ) }
            </Body1>
            <Button
              text={t( "RESTART-APP" )}
              onPress={( ) => RNRestart.restart()}
              className="my-5"
              level="focus"
            />
            <Heading2 className="mt-5">{ t( "Error" ) }</Heading2>
            <Body1>{ this.state.error.toString( ) }</Body1>
            <Body1>{ this.state.info?.componentStack }</Body1>
          </ScrollView>
        </SafeAreaView>
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
