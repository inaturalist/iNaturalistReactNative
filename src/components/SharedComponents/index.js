export { default as ActivityAnimation } from "./ActivityAnimation/ActivityAnimation";
export { default as ActivityCount } from "./ActivityCount/ActivityCount";
export { default as CommentsCount } from "./ActivityCount/CommentsCount";
export { default as IdentificationsCount } from "./ActivityCount/IdentificationsCount";
export { default as ActivityIndicator } from "./ActivityIndicator";
export { default as ButtonBar } from "./ButtonBar";
export { default as BackButton } from "./Buttons/BackButton";
export { default as Button } from "./Buttons/Button";
export { default as CloseButton } from "./Buttons/CloseButton";
export { default as EvidenceButton } from "./Buttons/EvidenceButton";
export { default as INatIconButton } from "./Buttons/INatIconButton";
export { default as RotatingINatIconButton } from "./Buttons/RotatingINatIconButton";
export { default as TransparentCircleButton } from "./Buttons/TransparentCircleButton";
export { default as Carousel } from "./Carousel";
export { default as CarouselDots } from "./CarouselDots";
export { default as Checkbox } from "./Checkbox";
export { default as CircleDots } from "./CircleDots";
export { default as ConfidenceInterval } from "./ConfidenceInterval";
export { default as DateDisplay } from "./DateDisplay";
export { default as DateTimePicker } from "./DateTimePicker";
export { default as DisplayTaxon } from "./DisplayTaxon";
export { default as DisplayTaxonName } from "./DisplayTaxonName";
export { default as Divider } from "./Divider/Divider";
export { default as CustomFlashList } from "./FlashList/CustomFlashList";
export { default as CustomRefreshControl } from "./FlashList/CustomRefreshControl";
export { default as FlashListEmptyWrapper } from "./FlashList/FlashListEmptyWrapper";
export { default as InfiniteScrollLoadingWheel } from "./FlashList/InfiniteScrollLoadingWheel";
export { default as FloatingActionBar } from "./FloatingActionBar";
export { default as HideView } from "./HideView";
export { default as IconicTaxonChooser } from "./IconicTaxonChooser";
export { default as IconicTaxonIcon } from "./IconicTaxonIcon";
export { default as INatIcon } from "./INatIcon";
export { default as InlineUser } from "./InlineUser/InlineUser";
export { default as InputField } from "./InputField";
export { default as KebabMenu } from "./KebabMenu";
export { default as KeyboardDismissableView } from "./KeyboardDismissableView";
export { default as LabelColonValue } from "./LabelColonValue";
export { default as DetailsMap } from "./Map/DetailsMap";
export { default as Map } from "./Map/Map";
export { default as MediaNavButtons } from "./MediaNavButtons";
export { default as Modal } from "./Modal";
export { default as Mortal } from "./Mortal";
export { default as ObservationLocation } from "./ObservationLocation";
export { default as ObsStatus } from "./ObsStatus";
export { default as OfflineNotice } from "./OfflineNotice";
export { default as OverlayHeader } from "./OverlayHeader";
export { default as OverviewCounts } from "./OverviewCounts";
export { default as PermissionGate } from "./PermissionGate";
export { default as PhotoCount } from "./PhotoCount";
export { default as QualityGradeStatus } from "./QualityGradeStatus/QualityGradeStatus";
export { default as RadioButtonRow } from "./RadioButtonRow";
export { default as ScrollableWithStickyHeader } from "./ScrollableWithStickyHeader";
export { default as ScrollViewWrapper } from "./ScrollViewWrapper";
export { default as SearchBar } from "./SearchBar";
export { default as BottomSheet } from "./Sheets/BottomSheet";
export { default as BottomSheetStandardBackdrop } from "./Sheets/BottomSheetStandardBackdrop";
export { default as PickerSheet } from "./Sheets/PickerSheet";
export { default as RadioButtonSheet } from "./Sheets/RadioButtonSheet";
export { default as TextInputSheet } from "./Sheets/TextInputSheet";
export { default as TextSheet } from "./Sheets/TextSheet";
export { default as WarningSheet } from "./Sheets/WarningSheet";
export { default as SimpleObservationLocation } from "./SimpleObservationLocation";
export { default as Tabs } from "./Tabs/Tabs";
export { default as TaxonResult } from "./TaxonResult";
export { default as TaxonSearch } from "./TaxonSearch";
export { default as Body1 } from "./Typography/Body1";
export { default as Body2 } from "./Typography/Body2";
export { default as Body3 } from "./Typography/Body3";
export { default as Body4 } from "./Typography/Body4";
export { default as Heading1 } from "./Typography/Heading1";
export { default as Heading2 } from "./Typography/Heading2";
export { default as Heading3 } from "./Typography/Heading3";
export { default as Heading4 } from "./Typography/Heading4";
export { default as Heading5 } from "./Typography/Heading5";
export { default as Heading6 } from "./Typography/Heading6";
export { default as List2 } from "./Typography/List2";
export { default as P } from "./Typography/P";
export { default as Subheading1 } from "./Typography/Subheading1";
export { default as Subheading2 } from "./Typography/Subheading2";
export { default as UnderlinedLink } from "./Typography/UnderlinedLink";
export { default as UploadProgressBar } from "./UploadProgressBar";
export { default as UploadStatus } from "./UploadStatus/UploadStatus";
export { default as UserIcon } from "./UserIcon";
export { default as UserText } from "./UserText";
export { default as ViewWrapper } from "./ViewWrapper";
