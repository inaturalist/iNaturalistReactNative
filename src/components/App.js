// @flow

import { useNavigation } from "@react-navigation/native";
import RootDrawerNavigator from "navigation/rootDrawerNavigator";
import type { Node } from "react";
import React, { useCallback } from "react";
import { log } from "sharedHelpers/logger";
import {
  useCurrentUser,
  usePerformance,
  useShare
} from "sharedHooks";
import { isDebugMode } from "sharedHooks/useDebugMode";

import AppStateListener from "./AppStateListener";
import useLinking from "./hooks/useLinking";
import NetworkService from "./NetworkService";
import StartupService from "./StartupService";

const logger = log.extend( "App" );

type SharedItem = {
  mimeType: string,
  data: string | Array<string>
};

const handleShare = ( navigation, item: ?SharedItem ) => {
  if ( !item ) {
    // user hasn't shared any items
    return;
  }

  const { mimeType, data } = item;

  if ( !mimeType && !data ) {
    // user hasn't shared any images
    return;
  }

  // show user a loading animation screen (like PhotoLibrary)
  // while observations are created
  navigation?.navigate( "PhotoSharing", { item } );
};

type Props = {
  // $FlowIgnore
  children?: unknown,
};

// this children prop is here for the sake of testing with jest
// normally we would never do this in code
const App = ( { children }: Props ): Node => {
  const navigation = useNavigation( );
  const { loadTime } = usePerformance( {
    screenName: "App"
  } );
  if ( isDebugMode( ) ) {
    logger.info( loadTime );
  }

  // attempting to make sure that navigation is only called once
  // for performance reasons
  const onShare = useCallback(
    item => handleShare( navigation, item ),
    [navigation]
  );

  const currentUser = useCurrentUser( );

  useLinking( currentUser );
  useShare( onShare );

  // this children prop is here for the sake of testing with jest
  // normally we would never do this in code
  return (
    <>
      <StartupService />
      <NetworkService />
      <AppStateListener />
      {children || <RootDrawerNavigator />}
    </>
  );
};

export default App;
