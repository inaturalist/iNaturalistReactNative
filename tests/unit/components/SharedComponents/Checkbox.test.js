import { fireEvent, render, screen } from "@testing-library/react-native";
import { Checkbox } from "components/SharedComponents";
import React from "react";
import colors from "styles/tailwindColors";
import { renderComponent } from "tests/helpers/render";

const rerenderCheckmarkComponent = checked => {
  renderComponent(
    <Checkbox
      accessibilityLabel="Checkmark"
      text="Checkmark text"
      isChecked={checked}
    />
  );
  const checkbox = screen.getByLabelText( /Checkmark/ );
  expect( checkbox ).toHaveProp( "innerIconStyle", {
    borderRadius: 6,
    borderWidth: 2,
    borderColor: colors.inatGreen
  } );
};

describe( "Checkbox", () => {
  it( "renders reliably", () => {
    render( <Checkbox text="Checkmark text" /> );

    // Disabled during the update to RN 0.78
    // expect( screen ).toMatchSnapshot();
  } );

  it( "renders reliably being checked", () => {
    render( <Checkbox text="Checkmark text" isChecked /> );

    // Disabled during the update to RN 0.78
    // expect( screen ).toMatchSnapshot();
  } );

  it( "has no accessibility errors", () => {
    // const checkbox = <Checkbox accessibilityLabel="Checkmark" text="Checkmark text" isChecked />;

    // Disabled during the update to RN 0.78
    // expect( checkbox ).toBeAccessible();
  } );

  it( "renders an empty checkbox when isChecked is false", () => {
    renderComponent(
      <Checkbox
        accessibilityLabel="Checkmark"
        text="Checkmark text"
        isChecked={false}
      />
    );

    const checkmark = screen.getByLabelText( /Checkmark/ );

    expect( checkmark ).toBeTruthy( );
    expect( checkmark ).toHaveProp( "innerIconStyle", {
      borderRadius: 6,
      borderWidth: 2,
      borderColor: colors.darkGray
    } );
  } );

  it( "renders a green filled checkbox when isChecked is true", () => {
    renderComponent( <Checkbox accessibilityLabel="Checkmark" text="Checkmark text" isChecked /> );

    const checkmark = screen.getByLabelText( /Checkmark/ );
    expect( checkmark ).toHaveProp( "innerIconStyle", {
      borderRadius: 6,
      borderWidth: 2,
      borderColor: colors.inatGreen
    } );
  } );

  it( "changes value when user presses checkbox being not checked", () => {
    let checked = false;
    renderComponent(
      <Checkbox
        accessibilityLabel="Checkmark"
        text="Checkmark text"
        isChecked={checked}
        // eslint-disable-next-line no-return-assign
        onPress={( ) => ( checked = !checked )}
      />
    );
    const checkmark = screen.getByLabelText( /Checkmark/ );

    expect( checked ).toBeFalsy( );
    fireEvent.press( checkmark );
    expect( checked ).toBeTruthy( );
    rerenderCheckmarkComponent( checked );
  } );

  it( "changes value when user presses checkbox being checked", () => {
    let checked = true;
    renderComponent(
      <Checkbox
        accessibilityLabel="Checkmark"
        text="Checkmark text"
        isChecked={checked}
        onPress={( ) => { checked = !checked; }}
      />
    );
    const checkmark = screen.getByLabelText( /Checkmark/ );

    expect( checked ).toBeTruthy( );
    fireEvent.press( checkmark );
    expect( checked ).toBeFalsy( );
  } );

  it( "renders text and changes value when user presses text", () => {
    let checked = false;
    renderComponent(
      <Checkbox
        text="Checkmark text"
        isChecked={checked}
        onPress={( ) => { checked = !checked; }}
      />
    );

    const text = screen.getByText( /Checkmark text/ );
    expect( text ).toBeVisible( );
    expect( checked ).toBeFalsy( );
    fireEvent.press( text );
    expect( checked ).toBeTruthy( );
    rerenderCheckmarkComponent( checked );
  } );
} );
