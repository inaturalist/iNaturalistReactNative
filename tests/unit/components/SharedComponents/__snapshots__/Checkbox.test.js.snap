// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders reliably 1`] = `
<View
  accessibilityRole="radio"
  accessibilityState={
    {
      "busy": undefined,
      "checked": false,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  disableBuiltInState={true}
  fillColor="#74AC00"
  focusable={true}
  iconComponent={null}
  iconStyle={
    {
      "borderRadius": 6,
    }
  }
  innerIconStyle={
    {
      "borderColor": "#454545",
      "borderRadius": 6,
      "borderWidth": 2,
    }
  }
  isChecked={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  size={25}
  style={
    [
      {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
    ]
  }
  textComponent={
    <ForwardRef
      className="ml-3 flex-shrink"
      component={[Function]}
    >
      Checkmark text
    </ForwardRef>
  }
  unfillColor="#ffffff"
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "borderRadius": 6,
        "height": 25,
        "justifyContent": "center",
        "transform": [
          {
            "scale": 1,
          },
        ],
        "width": 25,
      }
    }
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "borderColor": "#74AC00",
            "borderRadius": 12.5,
            "borderWidth": 1,
            "height": 25,
            "justifyContent": "center",
            "width": 25,
          },
          {
            "borderColor": "#454545",
            "borderRadius": 6,
            "borderWidth": 2,
          },
        ]
      }
    />
  </View>
  <Text
    maxFontSizeMultiplier={2}
    style={
      [
        {
          "color": "#454545",
        },
        {
          "fontFamily": "Lato-Regular",
        },
        [
          {
            "textAlign": "left",
          },
          [
            {
              "fontSize": 15,
              "lineHeight": 18,
            },
            {
              "fontFamily": "Lato-Regular",
            },
            [
              {
                "marginLeft": 12,
              },
              {
                "flexShrink": 1,
              },
            ],
          ],
        ],
      ]
    }
  >
    Checkmark text
  </Text>
</View>
`;

exports[`Checkbox renders reliably being checked 1`] = `
<View
  accessibilityRole="radio"
  accessibilityState={
    {
      "busy": undefined,
      "checked": true,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  disableBuiltInState={true}
  fillColor="#74AC00"
  focusable={true}
  iconComponent={
    <INatIcon
      color="#ffffff"
      name="checkmark"
      size={19}
    />
  }
  iconStyle={
    {
      "borderRadius": 6,
    }
  }
  innerIconStyle={
    {
      "borderColor": "#74AC00",
      "borderRadius": 6,
      "borderWidth": 2,
    }
  }
  isChecked={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  size={25}
  style={
    [
      {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
    ]
  }
  textComponent={
    <ForwardRef
      className="ml-3 flex-shrink"
      component={[Function]}
    >
      Checkmark text
    </ForwardRef>
  }
  unfillColor="#ffffff"
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#74AC00",
        "borderRadius": 6,
        "height": 25,
        "justifyContent": "center",
        "transform": [
          {
            "scale": 1,
          },
        ],
        "width": 25,
      }
    }
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "borderColor": "#74AC00",
            "borderRadius": 12.5,
            "borderWidth": 1,
            "height": 25,
            "justifyContent": "center",
            "width": 25,
          },
          {
            "borderColor": "#74AC00",
            "borderRadius": 6,
            "borderWidth": 2,
          },
        ]
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "#ffffff",
              "fontSize": 19,
            },
            null,
            {
              "fontFamily": "INatIcon",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        ï…§
      </Text>
    </View>
  </View>
  <Text
    maxFontSizeMultiplier={2}
    style={
      [
        {
          "color": "#454545",
        },
        {
          "fontFamily": "Lato-Regular",
        },
        [
          {
            "textAlign": "left",
          },
          [
            {
              "fontSize": 15,
              "lineHeight": 18,
            },
            {
              "fontFamily": "Lato-Regular",
            },
            [
              {
                "marginLeft": 12,
              },
              {
                "flexShrink": 1,
              },
            ],
          ],
        ],
      ]
    }
  >
    Checkmark text
  </Text>
</View>
`;
